### PR TITLE
Last update label, issue #10 ❔ 

### PR DESCRIPTION
**Pull request by developerfromjokela**
Issue #10 done in this pull request

Last update label can be turned off completely from settings if user doesn't like that feature

Here's an screenshot:
Screenshot_1596747205_censored

### CODE REVIEW

**Comment by ristomatti** [comment]
I'm quite confused with the terminology used in the variable names. Do "last update", "last movement", "last change" and "last event" and "latest parcel event" all refer to the same thing? If they do, they should all be use the same terminology. I'm going to pick "last event date" as the basis of the review comments. It sounds the most descriptive to me.

File: app/src/main/java/com/nitramite/adapters/CustomParcelsAdapterV2.java [hunk 1]
```diff
@@ -224,8 +227,16 @@ else if (parcelItems.get(position).getParcelCarrier().equals("99") && phase.equa
...
+        ParcelItem item = parcelItems.get(position);
+        parcelLastMovementStatusTV.setVisibility(!lastUpdate || (item.getLastEventDate() == null || parcelItems.get(position).getArchivedPackage()) ? View.GONE : View.VISIBLE);
         if (!parcelItems.get(position).getArchivedPackage()) {
             parcelUpdateStatusTV.setText(parcelItems.get(position).getParcelUpdateStatus());
+            if (lastUpdate) {
+                if (item.getLastEventDate() != null) {
+                    parcelUpdateStatusTV.setVisibility(View.GONE);
+                    parcelLastMovementStatusTV.setText(context.getString(R.string.last_change, item.getLastEventDate()));
+                }
+            }
```
Comment (ristomatti) [inline app/src/main/java/com/nitramite/adapters/CustomParcelsAdapterV2.java:239]: It's very difficult to follow what's going on here. First of all `parcelItems.get(position)` is set to a variable but it's then still used 3 times. If I understood correctly what this does, could this be refactored to something like below? It's more lines but arguably(?) easier to read.

```java
ParcelItem item = parcelItems.get(position);
boolean isArchived = item.getArchivedPackage();
boolean useLastEventDate = showLastEventDate && item.getLastEventDate() != null;
if (!isArchived) {
    parcelUpdateStatusTV.setText(item.getParcelUpdateStatus());
    if (useLastEventDate) {
        parcelUpdateStatusTV.setVisibility(View.GONE);
        parcelLastEventDateTV.setVisibility(View.VISIBLE);
        parcelLastEventDateTV.setText(context.getString(R.string.last_change, item.getLastEventDate()));
    } else {       
        parcelUpdateStatusTV.setVisibility(View.VISIBLE);
        parcelLastEventDateTV.setVisibility(View.GONE);
    }
}
```

File: app/src/main/java/com/nitramite/paketinseuranta/Constants.java [hunk 2]
```diff
@@ -21,6 +21,7 @@
     final static String SP_MAIN_MENU_LIST_LONG_TAP_ACTION = "SP_MAIN_MENU_LIST_LONG_TAP_ACTION";
     final static String SP_UPDATE_FAILED_FIRST = "SP_UPDATE_FAILED_FIRST";
     public final static String SP_THEME_SELECTION = "SP_THEME_SELECTION";
+    public final static String SP_PACKAGE_LAST_CHANGE = "SP_PACKAGE_LAST_CHANGE";
```
Comment (ristomatti) [inline app/src/main/java/com/nitramite/paketinseuranta/Constants.java:24]: Rename to `SP_SHOW_LAST_EVENT_DATE` for consistency and to make it clear this is a setting toggle (if I understood correctly).

File: app/src/main/java/com/nitramite/paketinseuranta/MainMenu.java [hunk 3]
```diff
@@ -97,6 +97,7 @@
     private View emptyView;
     private SharedPreferences sharedPreferences;
     private Boolean SP_UPDATE_FAILED_FIRST = false;
+    private boolean lastUpdate = false;
```
Comment (ristomatti) [inline app/src/main/java/com/nitramite/paketinseuranta/MainMenu.java:100]: Rename to `showLastEventDate` for consistency and make it clear it's a boolean and not the date string.

File: app/src/main/res/layout/parcel_item.xml [hunk 4]
```diff
@@ -153,6 +162,11 @@
 
                 </LinearLayout>
 
+                <TextView
+                    android:id="@+id/parcelLastMovementStatusTV"
```
Comment (ristomatti) [inline app/src/main/res/layout/parcel_item.xml:166]: Rename to `parcelLastEventDateTV` (if renamed also in `CustomParcelsAdapterV2.java` as I suggested).

File: app/src/main/res/xml/preferences.xml [hunk 5]
```diff
@@ -62,6 +62,12 @@
             android:summary="@string/preferences_update_failed_first_summary"
             android:title="@string/preferences_update_failed_first_title" />
 
+        <SwitchPreference
+            android:defaultValue="true"
+            android:key="SP_PACKAGE_LAST_CHANGE"
```
Comment (ristomatti) [inline app/src/main/res/xml/preferences.xml:67]: Rename to `SP_SHOW_LAST_EVENT_DATE` if renamed also in `Constants.java`.

**Comment by norkator** [comment]
I won't merge this but updated latest code from master here and corrected conflicts.

File: app/src/main/java/com/nitramite/adapters/CustomParcelsAdapterV2.java (hunk 1 above)
Comment (ristomatti) [reply app/src/main/java/com/nitramite/adapters/CustomParcelsAdapterV2.java:238]: @norkator comments on this? Would this achieve the same end result? I was not 100% sure as the particular code is quite "convoluted". :grimacing:

Comment (norkator) [reply app/src/main/java/com/nitramite/adapters/CustomParcelsAdapterV2.java:238]: > It's very difficult to follow what's going on here. First of all `parcelItems.get(position)` is set to a variable but it's then still used 3 times. If I understood correctly what this does, could this be refactored to something like below? It's more lines but arguably(?) easier to read.
> 
> ```java
> ParcelItem item = parcelItems.get(position);
> boolean isArchived = item.getArchivedPackage();
> boolean useLastEventDate = showLastEventDate && item.getLastEventDate() != null;
> if (!isArchived) {
>     parcelUpdateStatusTV.setText(item.getParcelUpdateStatus());
>     if (useLastEventDate) {
>         parcelUpdateStatusTV.setVisibility(View.GONE);
>         parcelLastEventDateTV.setVisibility(View.VISIBLE);
>         parcelLastEventDateTV.setText(context.getString(R.string.last_change, item.getLastEventDate()));
>     } else {       
>         parcelUpdateStatusTV.setVisibility(View.VISIBLE);
>         parcelLastEventDateTV.setVisibility(View.GONE);
>     }
> }
> ```

Hahaha this is my crap again. I need to fix this.

**Comment by norkator** [comment]
> I'm quite confused with the terminology used in the variable names. Do "last update", "last movement", "last change" and "last event" and "latest parcel event" all refer to the same thing? If they do, they should all be use the same terminology. I'm going to pick "last event date" as the basis of the review comments. It sounds the most descriptive to me.
> 
> Edit: @norkator please don't merge yet, I'm reviewing.

* Last update means when application updated / looked successfully parcel details from courier server
* Last change looks to be checking latest timestamp when parcel had something going on at carrier's side.

**Comment by norkator** [comment]
Now that I am thinking we should remove that old "when got successful answer from server timestamp" and replace it fully with last change (parcel event)

**Comment by developerfromjokela** [comment]
@norkator Fine for me. Can't we just set showing the last status update by default? I've put an setting there so user can switch it on or off

**Comment by developerfromjokela** [comment]
So set the switch preference by default to true? And if no status is available, then last refreshed date will be shown.

**Comment by norkator** [comment]
> @norkator Fine for me. Can't we just set showing the last status update by default? I've put an setting there so user can switch it on or off

Yes we can do that way too, it's already on by default: android:defaultValue="true"

**Comment by developerfromjokela** [comment]
Ok. I suggest let's keep it there, in my case I have a parcel that has no events but found in carrier's system. 

Let the app show last status update. When it's not available, last update will be shown.

What's your opinion @norkator?

**Comment by norkator** [comment]
Yeah, I start merging pull requests today to get ready for next update.

**Pull request merged**